controllers/kubeadmconfig_controller.go

```diff
@@ -300,6 +302,68 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 	return ctrl.Result{}, nil
 }
 
+// reconcileDiscovery ensure that joinConfiguration.Discovery is properly set for the joining node.
+// The implementation func respect user provided discovery configurations, but in case some of them are missing, a valid BootstrapToken object
+// is automatically injected into joinConfiguration.Discovery.
+// This allows to simplify configuration UX, by providing the option to delegate to CAPBK the configuration of kubeadm join discovery.
+func (r *KubeadmConfigReconciler) reconcileDiscovery(cluster *capiv1alpha2.Cluster, discovery *kubeadmv1beta1.Discovery, log logr.Logger) error {
+	// if config already contains a file discovery configuration, respect it without further validations
+	if discovery.File != nil {
+		return nil
+	}
+
+	// otherwise it is necessary to ensure token discovery is properly configured
+	if discovery.BootstrapToken == nil {
+		discovery.BootstrapToken = &kubeadmv1beta1.BootstrapTokenDiscovery{}
+	}
+
+	// if BootstrapToken already contains an APIServerEndpoint, respect it; otherwise inject the APIServerEndpoint endpoint defined in cluster status
+	//TODO(fp) might be we want to validate user provided APIServerEndpoint and warn/error if it doesn't match the api endpoint defined at cluster level
+	apiServerEndpoint := discovery.BootstrapToken.APIServerEndpoint
+	if apiServerEndpoint == "" {
+		if len(cluster.Status.APIEndpoints) == 0 {
+			return errors.New("Control plane already exists for the cluster but cluster.Status.APIEndpoints are not set")
+		}
+
+		// NB. CAPBK only uses the first APIServerEndpoint defined in cluster status if there are multiple defined.
+		apiServerEndpoint = fmt.Sprintf("https://%s:%d", cluster.Status.APIEndpoints[0].Host, cluster.Status.APIEndpoints[0].Port)
+		discovery.BootstrapToken.APIServerEndpoint = apiServerEndpoint
+		log.Info("Altering JoinConfiguration.Discovery.BootstrapToken", "APIServerEndpoint", apiServerEndpoint)
+	}
+
+	// if BootstrapToken already contains a token, respect it; otherwise create a new bootstrap token for the node to join
+	if discovery.BootstrapToken.Token == "" {
+		log.Info("Creating a bootstrap token in the remote cluster")
+		if r.remoteClientFactory == nil {
```

Comment:
We cannot put a remote client factory on the reconciler as the reconciler is a singleton that needs to work across many clusters.

Instead, I suggest creating a client when necessary during every reconcile.

Reply:
the factory is a function that gets a cluster as an input, so, it should be already designed to work across many clusters.

it is defined at reconciler level in order to get everything unit-testable